lib/auth.js

```diff
@@ -0,0 +1,454 @@
+'use strict';
+const _ = require('lodash'),
+  passport = require('passport'),
+  users = require('./services/users'),
+  db = require('./services/db'),
+  references = require('./services/references'),
+  session = require('express-session'),
+  flash = require('express-flash'),
+  handlebars = require('handlebars'),
+  basicAuth = require('basic-auth'),
+  fs = require('fs'),
+  path = require('path');
+
+/**
+ * determine if a route is protected
+ * protected routes are ?edit=true and any method other than GET
+ * @param {object} req
+ * @returns {boolean}
+ */
+function isProtectedRoute(req) {
+  return !!req.query.edit || req.method !== 'GET';
+}
+
+/**
+ * protect routes
+ * @param {object} site
+ * @returns {Function}
+ */
+function isAuthenticated(site) {
+  return function (req, res, next) {
+    if (req.isAuthenticated()) {
+      next(); // already logged in
+    } else if (req.get('Authorization')) {
+      // try to authenticate with api key
+      passport.authenticate('apikey', { session: false})(req, res, next);
+    } else {
+      req.session.returnTo = req.originalUrl; // redirect to this page after logging in
+      // otherwise redirect to login
+      res.redirect(`${site.path}/auth/login`);
+    }
+  };
+};
+
+/**
+ * get callback url for a site
+ * note: needs to add/not add initial slash, depending on the site path
+ * @param {object} site
+ * @param {string} provider
+ * @returns {string}
+ */
+function getCallbackUrl(site, provider) {
+  return site.path ?
+    `${references.uriToUrl(site.prefix, null, site.port)}/auth/${provider}/callback` :
+    `${references.uriToUrl(site.prefix, null, site.port)}auth/${provider}/callback`;
+}
+
+// serialize and deserialize users into the session
+/**
+ * call callback with user data
+ * @param {object} user
+ * @param {function} done
+ */
+function returnUser(user, done) {
+  done(null, user);
+}
+passport.serializeUser(returnUser);
+passport.deserializeUser(returnUser);
+
+/**
+ * create/authenticate against a clay user
+ * @param {object} properties to grab from provider and provider name itself
+ * @param {object} site
+ * @returns {Promise}
+ */
+function verify(properties, site) {
+  return function (req, token, tokenSecret, profile, done) { // eslint-disable-line
+    var username = _.get(profile, properties.username),
+      imageUrl = _.get(profile, properties.imageUrl),
+      name = _.get(profile, properties.name),
+      provider = properties.provider,
+      uid = site.prefix + '/users/' + users.encode(`${username.toLowerCase()}`, provider);
```

Comment:
I believe full UTF-8 works for keys, but yagni. if there's no username, it should error out because it means our code can't handle whatever provider they're authenticating against 😄